pyo3-macros-backend/src/attributes.rs

```diff
@@ -233,13 +233,28 @@ pub fn take_attributes(
 
 pub fn take_pyo3_options<T: Parse>(attrs: &mut Vec<syn::Attribute>) -> Result<Vec<T>> {
     let mut out = Vec::new();
-    take_attributes(attrs, |attr| {
-        if let Some(options) = get_pyo3_options(attr)? {
-            out.extend(options);
+    let mut allerr = Vec::new();
+    take_attributes(attrs, |attr| match get_pyo3_options(attr) {
+        Ok(result) => {
+            if let Some(options) = result {
+                out.extend(options);
+                Ok(true)
+            } else {
+                Ok(false)
+            }
+        }
+        Err(err) => {
+            allerr.extend(err);
             Ok(true)
-        } else {
-            Ok(false)
         }
     })?;
-    Ok(out)
+    if !allerr.is_empty() {
+        let mut error = allerr[0].clone();
+        for err in &allerr[1..] {
+            error.combine(err.clone());
+        }
+        Err(error)
+    } else {
+        Ok(out)
+    }
```

Comment:
Ah, this `error.combine()` function is extremely handy!

I think now that we know about this, to avoid the `allerr` vec we can probably have something like this:

```rust
struct ErrorCombiner(Option<syn::Error>);

impl ErrorCombiner {
    fn combine(&mut self, error: syn::Error) {
        if let Some(existing) = &mut self.0 {
            existing.combine(error);
        } else {
            self.0 = Some(error);
        }
    }

    fn ensure_empty(self) -> Result<()> {
        if let Some(error) = self.0 {
            Err(error)
        } else {
            None
        }
}
```

Above, instead of extending the `Vec` we would combine the error immediately, and then at the end of the function we could just check the combiner is empty:

```suggestion
    error_combiner.ensure_empty()?;
    Ok(out)
```

I think this pattern would be useful in a lot of places in the macro code!